hail/python/test/hailtop/test_fs.py

```diff
@@ -267,18 +269,60 @@ async def test_rmtree(filesystem):
     sema, fs, base = filesystem
 
     dir = f'{base}foo/'
+    subdir1 = f'{dir}foo/'
+    subdir1subdir1 = f'{subdir1}foo/'
+    subdir1subdir2 = f'{subdir1}bar/'
+    subdir1subdir3 = f'{subdir1}baz/'
+    subdir1subdir4_empty = f'{subdir1}qux/'
+    subdir2 = f'{dir}bar/'
+    subdir3 = f'{dir}baz/'
+    subdir4_empty = f'{dir}qux/'
 
     await fs.mkdir(dir)
     await fs.touch(f'{dir}a')
     await fs.touch(f'{dir}b')
 
+    await fs.mkdir(subdir1)
+    await fs.mkdir(subdir1subdir1)
+    await fs.mkdir(subdir1subdir2)
+    await fs.mkdir(subdir1subdir3)
+    await fs.mkdir(subdir1subdir4_empty)
+    await fs.mkdir(subdir2)
+    await fs.mkdir(subdir3)
+    await fs.mkdir(subdir4_empty)
+
+    sema = asyncio.Semaphore(100)
+    await bounded_gather2(sema, *[
+        functools.partial(fs.touch, f'{subdir}a{i:02}')
+        for subdir in [dir, subdir1, subdir2, subdir3, subdir1subdir1, subdir1subdir2, subdir1subdir3]
+        for i in range(30)])
+
     assert await fs.isdir(dir)
+    assert await fs.isdir(subdir1)
+    assert await fs.isdir(subdir1subdir1)
+    assert await fs.isdir(subdir1subdir2)
+    assert await fs.isdir(subdir1subdir3)
+    # subdir1subdir4_empty: in cloud fses, empty dirs do not exist and thus are not dirs
+    assert await fs.isdir(subdir2)
+    assert await fs.isdir(subdir3)
+    # subdir4_empty: in cloud fses, empty dirs do not exist and thus are not dirs
 
     await fs.rmtree(sema, dir)
 
     assert not await fs.isdir(dir)
 
 
+@pytest.mark.asyncio
+async def test_rmtree_empty_dir(filesystem):
```

Comment:
Do we need a test for a non-empty file that ends in a slash?

Reply:
Done.